src/marker.rs

```diff
@@ -522,26 +337,282 @@ impl<'py> Python<'py> {
+                    flag: Condvar::new(),
+                }
+            }
+
+            fn send_task(&self, task: Task) -> Option<Task> {
+                use MailboxInner::*;
+                let mut inner = self.inner.lock();
+                match &*inner {
+                    Empty => {
+                        *inner = Task(task);
+                        drop(inner);
+                        self.flag.notify_one();
+                        None
+                    }
+                    Abandoned => Some(task),
+                    Task(_) | Working | Done => unreachable!("sent task to active worker"),
+                }
+            }
+
+            fn recv_task(&self) -> Option<Task> {
+                use MailboxInner::*;
+                let mut inner = self.inner.lock();
+                loop {
+                    match &*inner {
+                        Empty | Done => {
+                            if self
+                                .flag
+                                .wait_for(&mut inner, Duration::from_secs(60))
+                                .timed_out()
+                            {
+                                *inner = Abandoned;
+                                return None;
+                            }
+                        }
+                        Task(_) => match replace(&mut *inner, Working) {
+                            Task(task) => return Some(task),
+                            _ => unreachable!(),
+                        },
+                        Working | Abandoned => {
+                            unreachable!("received task on active or exited worker")
+                        }
+                    }
+                }
+            }
+
+            fn signal_done(&self) {
+                use MailboxInner::*;
+                let mut inner = self.inner.lock();
+                match &*inner {
+                    Working => {
+                        *inner = Done;
+                        drop(inner);
+                        self.flag.notify_one();
+                    }
+                    Empty | Task(_) | Done | Abandoned => {
+                        unreachable!("signalled completion on inactive worker")
+                    }
+                }
+            }
+
+            fn await_done(&self) {
+                use MailboxInner::*;
+                let mut inner = self.inner.lock();
+                loop {
+                    match &*inner {
+                        Done => {
+                            *inner = Empty;
+                            return;
+                        }
+                        Task(_) | Working => self.flag.wait(&mut inner),
+                        Empty | Abandoned => {
+                            unreachable!("awaited completion from inactive worker")
+                        }
+                    }
+                }
+            }
+        }
+
+        static THREADS: Mutex<Vec<Arc<Mailbox>>> = const_mutex(Vec::new());
```

Comment:
There might be a possible simplification / optimisation here to use TLS (😂) to have a 1:1 relationship between a regular thread and a worker thread. That would remove the complexity around the mutex and the loop to find a non-dead mailbox.

The downside I guess is that there might potentially be more cold starts and higher peak thread count than if we shared workers between all threads. My gut tells me that if a thread is using `safe_allow_threads` it's likely to do so repeatedly in quick succession, so maybe there _is_ merit to this proposed simplification.

It might also come down to whether switching from a mutex to TLS has any meaningful impact on the microbenchmark?

Reply:
Also I guess an even more radical option would be to not have any global state in PyO3 and just expose this as an `AllowThreadsQueue` channel which users could use to manage their own worker threads. Personally I think I'm ok with this global state being here, but felt it worth tabling the option.

Reply:
> My gut tells me that if a thread is using safe_allow_threads it's likely to do so repeatedly in quick succession, so maybe there is merit to this proposed simplification.

I think so too as most Python still have a main thread that only branches out to schedule limited work onto thread pools and such. Will push that as a separate commit...

> Personally I think I'm ok with this global state being here, but felt it worth tabling the option.

I would like to avoid exposing this to user control to keep all our options to open to resolve this using a different implementation in the future.